altos-rust/port/cortex-m0/src/peripheral/usart/mod.rs

```diff
@@ -1,123 +1,305 @@
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
+ */
+
 mod control;
 mod defs;
 mod baudr;
 mod tdr;
+mod rdr;
 mod isr;
+mod icr;
+
+/// This module is the highest level in the Usart heirarchy for implementing
+/// the serial driver.
+///
+/// Configuration for each of the two Usart registers, and each of the registers
+/// that are offset from Usartx, and the public functions used to initialize,
+/// configure, and manipulate the bits for each register is defined in this file.
+///
+/// The functions here are used as wrappers that pass the call down through
+/// each necessary level (one or more), until the actual register is reached
+/// and is able to set the bits for itself accordingly.
+///
+/// This module is also responsible for initial setup of the Usart register
+/// (Either Usart1 or Usart2)
 
 use super::{Control, Register};
 use volatile::Volatile;
 use self::control::UsartControl;
 use self::baudr::BRR;
 use self::tdr::TDR;
+use self::rdr::RDR;
 use self::isr::ISR;
+use self::icr::ICR;
 use self::defs::*;
 use peripheral::{rcc, gpio};
+use interrupt;
 
 pub use self::control::{WordLength, Mode, Parity, StopLength, HardwareFlowControl};
 pub use self::baudr::BaudRate;
 
-#[derive(Copy, Clone)]
+pub const USART2_TX_BUFFER_FULL_CHAN: usize = 43;
+pub const USART2_TC_CHAN: usize = 43 * 2;
+
+/// STM32F0 has two Usart registers available.
+#[derive(Copy, Clone, Debug)]
 pub enum UsartX {
     Usart1,
     Usart2,
 }
 
-#[derive(Copy, Clone)]
+/// Defines the memory address for the Usart and each variable needed
+/// to access the registers that comprise the Usart.
+/* mem_addr: Memory address of the Usart
+ * control: Collection of control registers
+ * baud: Baud rate register
+ * tdr: Transmit data register
+ * rdr: Read data register
+ * isr: Interrupt service register
+ * icr: Interrupt clear register
+ */
+#[derive(Copy, Clone, Debug)]
 pub struct Usart {
     mem_addr: *const u32,
     control: UsartControl,
     baud: BRR,
     tdr: TDR,
+    rdr: RDR,
     isr: ISR,
+    icr: ICR,
 }
 
+/// Wraps the Usart memory address in order to make it volatile.
 impl Control for Usart {
     unsafe fn mem_addr(&self) -> Volatile<u32> {
         Volatile::new(self.mem_addr as *const u32)
     }
 }
 
+/// Function implementations for the Usart struct. These are the wrapper
+/// functions that can be called on a Usart object in order to set
+/// specific bits in each of the registers contained within the Usart struct.
+/// Each function calls the appropriate function associated with the
+/// register needed to set the correct bits.
 impl Usart {
+
+    /// Creates a new Usart. Pattern matches the argument passed in, and
+    /// creates and initializes either a new Usart1 or a new Usart2.
+    /// The base address of the Usart being used is passed into each 'new` call
+    /// the proper address with offset can be calculate for the associated
+    /// register.
```

Comment:
We don't really want to be talking about implementation details for our doc comments, doc comments are meant for usage, so we just want to explain what the function does, not how it does it.